src/hardware.rs

```diff
@@ -0,0 +1,208 @@
+use std::ops::{Index, IndexMut};
+
+pub enum HardwareError {
+    ErrorDecodingRegister,
+    ErrorDecodingFlag,
+    InvalidInstruction,
+    InvalidICode,
+}
+pub enum Register {
+    R0,
+    R1,
+    R2,
+    R3,
+    R4,
+    R5,
+    R6,
+    R7,
+    PC, /* program counter */
+    COND,
+}
+
+impl Register {
+    pub fn from_u16(value: u16) -> Result<Self, HardwareError> {
+        match value {
+            0 => Ok(Self::R0),
+            1 => Ok(Self::R1),
+            2 => Ok(Self::R2),
+            3 => Ok(Self::R3),
+            4 => Ok(Self::R4),
+            5 => Ok(Self::R5),
+            6 => Ok(Self::R6),
+            7 => Ok(Self::R7),
+            8 => Ok(Self::PC),
+            9 => Ok(Self::COND),
+            _ => {
+                Err(HardwareError::ErrorDecodingRegister) //Invalid Register
```

Comment:
To provide more context about the error, it would be helpful for `ErrorDecodingRegister(u16)` to display the value that triggered the error. This will make debugging easier when unexpected issues arise.

Apply this to all the code errors